src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs

```diff
@@ -0,0 +1,794 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.Arm;
+using System.Runtime.Intrinsics.X86;
+using System.Text;
+using System.Text.Unicode;
+using static System.Buffers.Text.Base64;
+
+namespace System.Buffers.Text
+{
+    // AVX2 and Vector128 version based on https://github.com/gfoidl/Base64/blob/5383320e28cac6c7ac6f86502fb05d23a048a21d/source/gfoidl.Base64/Internal/Encodings/Base64UrlEncoding.cs
+
+    public static partial class Base64Url
+    {
+        private const int MaxStackallocThreshold = 256;
+
+        /// <summary>
+        /// Returns the maximum length (in bytes) of the result if you were to decode base 64 encoded text within a byte span of size <paramref name="base64Length"/>.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="base64Length"/> is less than 0.
+        /// </exception>
+        public static int GetMaxDecodedLength(int base64Length)
+        {
+            ArgumentOutOfRangeException.ThrowIfNegative(base64Length);
+
+            (uint whole, uint remainder) = uint.DivRem((uint)base64Length, 4);
+
+            return (int)(whole * 3 + (remainder > 0 ? remainder - 1 : 0));
+        }
+
+        /// <summary>
+        /// Decode the span of UTF-8 encoded text represented as Base64Url into binary data.
+        /// </summary>
+        /// <param name="source">The input span which contains UTF-8 encoded text in Base64Url that needs to be decoded.</param>
+        /// <param name="destination">The output span which contains the result of the operation, i.e. the decoded binary data.</param>
+        /// <param name="bytesConsumed">The number of input bytes consumed during the operation. This can be used to slice the input for subsequent calls, if necessary.</param>
+        /// <param name="bytesWritten">The number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary.</param>
```

Comment:
https://github.com/dotnet/dotnet-api-docs/wiki/Parameters%3A-Methods-and-constructors, out parameters.


```suggestion
        /// <param name="bytesConsumed">When this method returns, contains the number of input bytes consumed during the operation. This can be used to slice the input for subsequent calls, if necessary. This parameter is treated as uninitialized.</param>
        /// <param name="bytesWritten">When this method returns, contains the number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary. This parameter is treated as uninitialized.</param>
```